File: packages/youtube-player/android/src/main/java/io/capawesome/capacitorjs/plugins/youtubeplayer/YoutubePlayerHelper.java
package io.capawesome.capacitorjs.plugins.youtubeplayer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.PluginCall;
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants;
import io.capawesome.capacitorjs.plugins.youtubeplayer.classes.CustomExceptions;

public class YoutubePlayerHelper {

    @NonNull
    public static String getPlayerIdFromCall(@NonNull PluginCall call) throws Exception {
        String id = call.getString("id");
        if (id == null) {
            throw CustomExceptions.PLAYER_ID_MISSING;
        }
        return id;
    }

    @NonNull
    public static PlayerConstants.PlaybackRate mapPlaybackRate(double rate) throws Exception {
        if (rate == 0.25) {
            return PlayerConstants.PlaybackRate.RATE_0_25;
        } else if (rate == 0.5) {
            return PlayerConstants.PlaybackRate.RATE_0_5;
        } else if (rate == 0.75) {
            return PlayerConstants.PlaybackRate.RATE_0_75;
        } else if (rate == 1) {
            return PlayerConstants.PlaybackRate.RATE_1;
        } else if (rate == 1.25) {
            return PlayerConstants.PlaybackRate.RATE_1_25;
        } else if (rate == 1.5) {
            return PlayerConstants.PlaybackRate.RATE_1_5;
        } else if (rate == 1.75) {
            return PlayerConstants.PlaybackRate.RATE_1_75;
        } else if (rate == 2) {
            return PlayerConstants.PlaybackRate.RATE_2;
        } else {
            throw CustomExceptions.RATE_INVALID;
        }
    }

    public static double mapPlaybackRateToDouble(@NonNull PlayerConstants.PlaybackRate playbackRate) {
        switch (playbackRate) {
            case RATE_0_25:
                return 0.25;
            case RATE_0_5:
                return 0.5;
            case RATE_0_75:
                return 0.75;
            case RATE_1_25:
                return 1.25;
            case RATE_1_5:
                return 1.5;
            case RATE_1_75:
                return 1.75;
            case RATE_2:
                return 2;
            default:
                return 1;
        }
    }

    @NonNull
    public static String mapPlayerError(@NonNull PlayerConstants.PlayerError error) {
        switch (error) {
            case INVALID_PARAMETER_IN_REQUEST:
                return "invalid-parameter";
            case HTML_5_PLAYER:
                return "html5-error";
            case VIDEO_NOT_FOUND:
                return "video-not-found";
            case VIDEO_NOT_PLAYABLE_IN_EMBEDDED_PLAYER:
                return "not-embeddable";
            default:
                return "unknown";
        }
    }

    @Nullable
    public static String mapPlayerState(@NonNull PlayerConstants.PlayerState state) {
        switch (state) {
            case UNSTARTED:
                return "unstarted";
            case ENDED:
                return "ended";
            case PLAYING:
                return "playing";
            case PAUSED:
                return "paused";
            case BUFFERING:
                return "buffering";
            case VIDEO_CUED:
                return "cued";
            default:
                return null;
        }
    }
}

File: packages/youtube-player/ios/Plugin/YoutubePlayerHelper.swift
import Capacitor
import Foundation
import YoutubePlayerView

public class YoutubePlayerHelper {
    static let allowedPlaybackRates: [Double] = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]

    static func getPlayerIdFromCall(_ call: CAPPluginCall) throws -> String {
        guard let id = call.getString("id") else {
            throw CustomError.playerIdMissing
        }
        return id
    }

    static func mapPlayerError(_ error: Error) -> String {
        guard let playerError = error as? YoutubePlayerError else {
            return "unknown"
        }
        switch playerError {
        case .invalidParam:
            return "invalid-parameter"
        case .HTML5Error:
            return "html5-error"
        case .videoNotFound:
            return "video-not-found"
        case .notEmbeddable:
            return "not-embeddable"
        case .unknown:
            return "unknown"
        }
    }

    static func mapPlayerState(_ state: YoutubePlayerState) -> String? {
        switch state {
        case .unstarted:
            return "unstarted"
        case .ended:
            return "ended"
        case .playing:
            return "playing"
        case .paused:
            return "paused"
        case .buffering:
            return "buffering"
        case .queued:
            return "cued"
        case .unknown:
            return nil
        }
    }
}
